File: C/2044-count-number-of-maximum-bitwise-or-subsets/main.go
package main

import "fmt"

func countMaxOrSubsets(nums []int) int {
	m := map[int]int{}
	m[0] = 1
	sum := 0
	for _, n := range nums {
		sum |= n
		b := map[int]int{}
		for k, v := range m {
			nn := k | n
			b[nn] += v
		}
		for k, v := range b {
			m[k] += v
		}
	}

	return m[sum]
}

func main() {
	fmt.Println(countMaxOrSubsets([]int{3, 1}))
	fmt.Println(countMaxOrSubsets([]int{2, 2, 2}))

}

File: C/2044-count-number-of-maximum-bitwise-or-subsets/main.rs
use std::collections::HashMap;

impl Solution {
  pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
    let mut m: HashMap<i32, i32> = HashMap::new();
    m.insert(0, 1);
    let mut sum: i32 = 0;
    nums.iter().for_each(|&n| {
      sum |= n;
      let mut b: HashMap<i32, i32> = HashMap::new();
      for (k, v) in m.iter() {
        b.entry(k | n).and_modify(|x| *x += v).or_insert(*v);
      }
      for (k, v) in b.iter() {
        m.entry(*k).and_modify(|x| *x += v).or_insert(*v);
      }
    });
    *m.get(&sum).unwrap()
  }
}
